Documentation (#38)
Arcadia-Science/metagenomics: docs/usage.md

# Arcadia-Science/metagenomics Usage

This page provides documentation on how to use the Arcadia-Science/metagenomics pipeline.

## Input specifications

> This pipeline processes metagenomic reads in FASTQ format from either Illumina or Nanopore technologies. **Important note**: This pipeline separately processes Illumina or Nanopore metagenomes and does not perform hybrid assembly, polishing of Nanopore assemblies with Illumina reads, or scaffolding with Nanopore reads.

Reviewer: I think this could be on the main README somewhere as well if you can find a place that it fits well. I think this is probably a pretty important statement

> For Nanopore reads, the pipeline takes as direct input a single fastq file and does not perform demultiplexing or basecalling, so it expects that those steps should be done prior to feeding into the pipeline and aggregating into a single fastq file.

Reviewer: ```suggestion
For Nanopore reads, the pipeline takes as direct input a single fastq file and does not perform demultiplexing or basecalling, so it expects that those steps have been done prior to feeding into the pipeline and aggregating into a single fastq file.
```

The workflow takes a CSV samplesheet listing the sample names and paths to the fastq files as direct input. It does not take the path to the directory of the fastq files, you must list them in a samplesheet. The CSV must contain the columns:

`sample, fastq_1, fastq_2`

> Neither the `sample` or `fastq*` columns may contain spaces, and the `fastq` columns contains the path(s) to your fastq file(s).

Reviewer 2: Why is that? Especially for `sample`?

Author: I guess it could be ok if sample had spaces, because names of files that incorporate pieces of the sample name come from the fastqs or downstream assemblies. I think it's a good rule of thumb though

If you are inputting Illumina reads, the samplesheet looks like the following:

```
sample,fastq_1,fastq_2
comm_1,https://github.com/Arcadia-Science/test-datasets/raw/main/metagenomics/illumina/comm_1_subsampled_1.fq.gz,https://github.com/Arcadia-Science/test-datasets/raw/main/metagenomics/illumina/comm_1_subsampled_2.fq.gz
vir_1,https://github.com/Arcadia-Science/test-datasets/raw/main/metagenomics/illumina/vir_1_subsampled_1.fq.gz,https://github.com/Arcadia-Science/test-datasets/raw/main/metagenomics/illumina/vir_1_subsampled_2.fq.gz
```

If you are inputting Nanopore reads, the samplesheet looks like the following:

```
sample,fastq_1,fastq_2
om,https://github.com/Arcadia-Science/test-datasets/raw/main/metagenomics/ont/om_subset_reads.fq.gz,
el,https://github.com/Arcadia-Science/test-datasets/raw/main/metagenomics/ont/el_subset_reads.fq.gz,
```

Note that even for Nanopore reads which the input is in a single fastq file, still include the third column for `fastq_2` - the pipeline for processing Nanopore reads will ignore this and process your Nanopore reads that are in a single fastq file.

## Running the pipeline

A typical command for running the pipeline looks like:

> ` nextflow run Arcadia-Science/metagenomics --input <SAMPLESHEET.csv> --outdir <OUTDIR> --platform <illumina|nanopore> -profile docker`

Reviewer 2: I like this example of platform more than what you use in the main readme!

Reviewer: +1

Reviewer: also WAIT! if I have nextflow installed, will this just work without cloning the repo??? if that works that is very cool. can you specify which version or commit or release to pull?

Author: I thought with the `-r` flag you could specify the release to launch from without cloning the repo

> This will launch the pipeline with the `docker` configuration profile. See below for more information about profiles. The options for `--platform` are either `illumina` or `nanopore` depending on your sequencing file input. Therefore, you cannot run both Illumina and Nanopore sequencing files in the same run - instead submit separate runs. The pipeline will create the following files and directories in your working directory:

Reviewer: can I have clashing sample names if I sequenced a sample with both illumina and nanopore?

```
work                # Directory containing the nextflow working files
<OUTDIR>            # Finished results in specified location (defined with --outdir)
.nextflow_log       # Log file from Nextflow
# Other nextflow hidden files, eg. history of pipeline runs and old logs.
```

### Updating the pipeline

> When you run the above command, Nextflow automatically pulls the pipeline code from GitHub and stores it as a cached version. When running the pipeline after this, it will always use the cached version if available - even if the pipeline has been updated since. To make sure you are running the latest version of the pipeline, regularly update the cached version with:

Reviewer 2: I feel like I reviewed this before, but maybe I only thought about it and haven't commented on it. I think this is an overly complicated description of what's happening. Nextflow pulls the repo locally and uses the local version. So if you need to update the pipeline code, pull the latest code. No need to mention "caches" etc.

Author: you've reviewed this in the hifi2genome workflow docs, it is the same pulled from there. I can modify and adjust in both places

`nextflow pull Arcadia-Science/metagenomics`

## Pipeline arguments

Below lists several arguments to configure core parts of how Nextflow is run and pipeline-specific arguments you can modify.

### Core Nextflow arguments

> These options are part of Nextflow and use a single hyphen.

Reviewer: can you run something like `nextflow -h` to see all of the available parameters? below you describe a subset of useful ones, only some of which are required right?

#### `-profile`

> Use this parameter to choose a configuration profile. Profiles can give configuration presets for different compute environments. Several generic profiles are bundled with the pipeline which instruct the pieline to use software packaged using different methods (Docker, Singularity, Conda, etc.) When using Biocontainers, most of the software packaging methods pull Docker containers from quay.io except for Singularity which directly downloads Singularity images via https hosted by the Galaxy project and Conda which downloads and installs software locally from Bioconda.

Reviewer: ```suggestion
Use this parameter to choose a configuration profile. Profiles can give configuration presets for different compute environments. Several generic profiles are bundled with the pipeline which instruct the pieline to use software packaged using different methods (Docker, Singularity, Conda, etc.). When using Biocontainers, most of the software packaging methods pull Docker containers from quay.io except for Singularity which directly downloads Singularity images via https hosted by the Galaxy project and Conda which downloads and installs software locally from Bioconda.
```

Reviewer: also this second sentence is a lot...I don't think "Biocontainers" is ever something that is exposed to an end user. I think focusing on these details from the docker, singularity, and conda perspectives only might help. And also maybe splitting it into three sentences.
When using the docker profile, XXX
when using the singularity profile, XXX
when using the conda profile, XXX

> We highly recommend the user of Docker or Singularity containers for full pipeline reproducibility, and have personally tested using Docker containers extensively within Arcadia. Other generic profiles besides the above listed are available, but they are untested and you can use at your own discretion.

Reviewer: ```suggestion
We highly recommend the use of Docker or Singularity containers for full pipeline repeatability and reproducibility, and have tested using Docker containers extensively within Arcadia. Other generic profiles besides the above listed are available, but they are untested and you can use at your own discretion.
```

If `-profile` is not specified, the pipeline will run locally and expect all software to be installed and available on the `PATH` which is not recommended.

- `docker`
  - A generic configuration profile to be used with Docker
- `singularity`
  - A generic configuration profile to be used with Singularity
- `conda`
  - A generic configuration profile to be used with Conda. Please only use Conda as a last resort, such as when it is not possible to run with Docker, Singularity, or the other preset configurations.
- `test`, `test_full`
  - Profiles with a complete configuration for automated testing. This includes links to test data and need no other parameters

#### `-resume`

> Specify this when restarting the pipeline from a previously failed run, updated part of the pipeline, or additional samples. Nextflow will use cached results from any pipeline steps where the inputs are the same, continuing from where it got to previously. For the input to be considered the same, not only the names of the files must be identical but all the files' contents as well.

Reviewer: and the code that generates those files right?

#### `-c`

> Specify the path to a specific config file.

Reviewer: what does the config file have in it? do I have to specify this? If i don't, what file is the default?

Author: I believe this is if you have to provide an extra config for like an institution HPC or configs that aren't covered by default

#### Running in the background

> Nextflow handles job submissions and supervises running jobs. The Nextflow process must run until the pipeline is finished. The Nextflow `-bg` flag launches Nextflow in the background, detached from your terminal so that the workflow does not stop if you log out of your session. The logs are saved to a file. Alternatively you can use `screen`/`tmux` or a similar tool to create a detached session which you can log back into at a later time.

Reviewer: > The Nextflow process must run until the pipeline is finished.

Is that sentence true? is it necessary to include it here?
